Add new game: 2048

Jazeem opened the pull request:
Clone of https://github.com/gabrielecirulli/2048

lanctot commented:
Very nice, thanks @Jazeem !!

I'm on vacation at the moment but will take a look when I'm back next week.

Jazeem commented:
Sure @lanctot 

I'm curious to see if any algorithms implemented in open_spiel would be able to reach 2048. I tried the default mcts implementation, but it was not faring better than a random player.

lanctot commented:
Hi @Jazeem, 

Hmm.. MCTS performing similarly to random-- even for MCTS with random rollouts-- I'm surprised. I suspect a bug in the game implementation. How many simulations did MCTS have? And how long is the maximum length of each game?

Other algorithms: you could also try some RL algorithms like DQN. Or, expectimax with a heuristic evaluation function.

lanctot commented:
Oh wait, depends on what you decided for the reward structure. If there are no intermediate rewards then I can see why it would be super hard for vanilla MCTS but still I would expect it to be better than random.

Jazeem commented:
@lanctot yeah you are correct. I think it's because there's no intermediate rewards. 

When I ran the simulation over 100 games, MCTS and random couldn't win once. 
But when I reduced the winning condition to tile 128 instead of 2048, MCTS could win 100 games vs 48 games for random.

steventrouble commented:
I suspect the terminal condition and reward function still need some tweaking. 2048 doesn't just end when the player reaches 2048, it keeps going forever. The score isn't win/lose, it's a large number that increases based on the number and size of the merged tiles.

Current SOTA algorithms can get scores of ~500k, which is far beyond reaching 2048.

Jazeem commented:
Hi @steventrouble 

I was experimenting around the terminal condition and reward function. My initial assumption was that setting the game to end at 2048 might make it easier to solve.

I've since then modified the reward system to give a reward for every new tile unlocked. But like you suggested it might be better to keep the game forever like the original and modify Rewards() as score from current action and Returns() as total score.

lanctot commented:
> Hi @steventrouble
> 
> I was experimenting around the terminal condition and reward function. My initial assumption was that setting the game to end at 2048 might make it easier to solve.

I recommend we make this as consistent with the original scoring (or with other RL implementations) as possible. 

> I've since then modified the reward system to give a reward for every new tile unlocked. But like you suggested it might be better to keep the game forever like the original and modify Rewards() as score from current action and Returns() as total score.

I think this should be configurable by a game parameter. All games in OpenSpiel are episodic: the tests assume this, so they should end in finite time (e.g. the game simulation tests can't go on forever) . I recommend you make a flag like "max_steps" or "max_game_length" parameter that plays 2048 for a finite number of steps *or* a flag that ends the episode after getting 2048 as the default, and then supporting the infinite game as a non-default parameter value.

lanctot commented on open_spiel/games/2048.cc:
```diff
@@ -0,0 +1,383 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
```
Can you add a comment block here that describes the game?

Also, it should include (i) a link to the Wikipedia description, and (ii) a link to the code this was based on (https://github.com/gabrielecirulli/2048) depending on how much you used.

Jazeem replied:
Have added this in 2048.h, should I move it here?

lanctot commented:
> Clone of https://github.com/gabrielecirulli/2048

Can you clarify how much of the code in here is based on this implementation? If you actually copied code from that repos, we likely have to put their copyright notice in the file before we can import. (And also it'd be a nice courtesy to tag the authors to let them know, they might have some comments / preferences).

lanctot commented on open_spiel/games/2048.cc:
```diff
@@ -0,0 +1,383 @@
+#include "open_spiel/utils/tensor_view.h"
+
+namespace open_spiel {
+namespace two_zero_four_eight {
```
Why not `twenty_fortyeight`, just curious? I've mostly heard people refer to it like that.

Jazeem replied:
That's interesting. I've only heard it being referred as two_zero_four_eight. I can change it if twenty_fortyeight is more common

lanctot commented on open_spiel/games/2048.cc:
```diff
@@ -0,0 +1,383 @@
+constexpr int kMoveRight = 1;
+constexpr int kMoveDown = 2;
+constexpr int kMoveLeft = 3;
+const std::vector<Action> kPlayerActions
```
Style thing: this can't be global because it's not trivially destructible.

Use a function instead: 
```
inline const std::vector<Action> kPlayerActions() {
  return {kMoveUp, kMoveRight, kMoveDown, kMoveLeft};
```

Jazeem replied:
Have changed this

lanctot commented on open_spiel/games/2048.cc:
```diff
@@ -0,0 +1,383 @@
+
+TwoZeroFourEightState::TwoZeroFourEightState(std::shared_ptr<const Game> game)
+    : State(game) {
+  board_ = std::vector<Tile>(kDefaultRows * kDefaultColumns, Tile(0, false));
```
initialize board also in the initializer list:

```
  : State(game),
    board(kDefaultRows * kDefaultColumns, Tile(0, false)) {}
```

Jazeem replied:
Have changed this

lanctot commented on open_spiel/games/2048.cc:
```diff
@@ -0,0 +1,383 @@
+  for (int x = 0; x < kDefaultRows; x++) {
+    for (int y = 0; y < kDefaultColumns; y++) {
+      SetBoard(x, y, Tile(board_seq[x * kDefaultRows + y], false));
+    }
```
use `r` and `c` rather than `x` and `y` as better names

Jazeem replied:
Have changed this

lanctot commented on open_spiel/games/2048.h:
```diff
@@ -0,0 +1,144 @@
+    * kChanceTiles.size();
+
+struct Coordinate {
+  int x, y;
```
either use better names (row, column) or add a comment to explain which is representing which

Jazeem replied:
Have changed this

lanctot commented on open_spiel/games/2048.h:
```diff
@@ -0,0 +1,144 @@
+    return std::unique_ptr<State>(new TwoZeroFourEightState(*this));
+  }
+  void UndoAction(Player player, Action action) override;
+  bool InBounds(int row, int column) const;
```
It's good practice to separate the methods below by adding a comment here that says the ones below are outside the core API (to distinguish the ones implementing the core API)

Jazeem replied:
Have changed this

lanctot commented on open_spiel/games/2048.h:
```diff
@@ -0,0 +1,144 @@
+  }
+  void UndoAction(Player player, Action action) override;
+  bool InBounds(int row, int column) const;
+  void SetCustomBoard(const std::vector<int> board_seq);
```
const std::vector<int>& board_seq

(use a reference)

Jazeem replied:
Have changed this

lanctot commented on open_spiel/games/2048.h:
```diff
@@ -0,0 +1,144 @@
+  Tile BoardAt(int row, int column) const {
+    return board_[row * kDefaultColumns + column];
+  }
+  std::vector<Action> LegalActions() const override;
```
move these up to group them with the other core API methods

Jazeem replied:
Have changed this

lanctot commented on open_spiel/games/2048.h:
```diff
@@ -0,0 +1,144 @@
+    return {kDefaultRows, kDefaultColumns};
+  }
+  // There is arbitrarily chosen number to ensure the game is finite.
+  int MaxGameLength() const override { return 1000; }
```
Make this configurable via a game parameter

Jazeem replied:
Have changed this

lanctot commented on open_spiel/games/2048.h:
```diff
@@ -0,0 +1,144 @@
+    return absl::make_unique<TwoZeroFourEightState>(shared_from_this());
+  }
+  int NumPlayers() const override { return kNumPlayers; }
+  double MinUtility() const override { return -1; }
```
These don't seem correct for this game?

(min utility should be 0, right? Max utility will depend on a few things)

Jazeem replied:
Have changed this

lanctot commented on open_spiel/games/2048.cc:
```diff
@@ -0,0 +1,383 @@
+  switch (action_id) {
+    case kMoveUp:
+      return "Up";
+      break;
```
Don't need the break here because you have a return

Jazeem replied:
My bad. Fixed this.

lanctot commented on open_spiel/games/2048.cc:
```diff
@@ -0,0 +1,383 @@
+  ActionsAndProbs action_and_probs;
+  int count = AvailableCellCount();
+  if (count == 0) {
+    action_and_probs.reserve(1);
```
this should work and is simpler: `return {{kNoCellAvailableAction, 1.0}};`

Jazeem replied:
Have changed this

lanctot commented on open_spiel/games/2048.cc:
```diff
@@ -0,0 +1,383 @@
+  if (IsTerminal()) {
+    if (Reached2048()) {
+      return {1.0};
+    }
```
Ah yes, I see now, you're treating it as a win/loss game. In this case, we should probably be consistent with how the RL community has treated this environment.

Jazeem replied:
Have changed it from a win/loss game to match the scoring system of the original game

Jazeem commented:
> I think this should be configurable by a game parameter. All games in OpenSpiel are episodic: the tests assume this, so they should end in finite time (e.g. the game simulation tests can't go on forever) . I recommend you make a flag like "max_steps" or "max_game_length" parameter that plays 2048 for a finite number of steps or a flag that ends the episode after getting 2048 as the default, and then supporting the infinite game as a non-default parameter value.

I've modified Returns() and Rewards() to be consistent with the original 2048 scoring. Have also introduced game parameters max_game_length and max_score (Currently have set them to INT_MAX, if this can cause any for any tests, training algorithms I can change this)

Jazeem commented:
> Clone of https://github.com/gabrielecirulli/2048
> Can you clarify how much of the code in here is based on this implementation? If you actually copied code from that repos, we likely have to put their copyright notice in the file before we can import. (And also it'd be a nice courtesy to tag the authors to let them know, they might have some comments / preferences).

Haven't actually copied any code since the original is written in Javascript and also had to make some changes to fit the LegalActions(), LegalChanceOutcomes(), DoApplyAction() architecture of open_spiel. 

But I did use some of the same functions in the original repo and rewrote it in cpp. Those are BuildTraversals(), FindFarthestPosition(), TileMatchesAvailable(), PrepareTiles()

Also DoApplyAction() was written by looking to the move() function in the original repo

Jazeem commented:
> (And also it'd be a nice courtesy to tag the authors to let them know, they might have some comments / preferences).

That'd be great!

tewalds commented on open_spiel/games/2048.cc:
```diff
@@ -0,0 +1,382 @@
+    Coordinate direction_diff = GetVector(direction);
+    r += direction_diff.row;
+    c += direction_diff.column;
+  } while (WithinBounds(r, c) && CellAvailable(r, c));
```
If you want to be fancy, you could use Coordinate in more places, like CellAvailable(coord) and have Coordinate::WithinBounds

tewalds commented on open_spiel/games/2048.cc:
```diff
@@ -0,0 +1,382 @@
+    }
+}
+
+std::vector<Coordinate> TwoZeroFourEightState
```
Since this will always return two, it could be a std::array<Coordinate, 2>

Jazeem replied:
Changed this

tewalds commented on open_spiel/games/2048.cc:
```diff
@@ -0,0 +1,382 @@
+  Coordinate prev = Coordinate(r, c);
+  do {
+    prev = Coordinate(r, c);
+    Coordinate direction_diff = GetVector(direction);
```
Given the direction doesn't change, the vector is constant. The compiler might figure that out, but you could force it by hoisting it out of the loop. Making it constexpr would also do it.

Jazeem replied:
Changed this

tewalds commented on open_spiel/games/2048.cc:
```diff
@@ -0,0 +1,382 @@
+        Coordinate next_pos = positions[1];
+        int next_cell = GetCellContent(next_pos.row, next_pos.column);
+        if (next_cell > 0 && next_cell == tile
+            && !BoardAt(next_pos.row, next_pos.column).is_merged) {
```
This would be cleaner if BoardAt took a Coordinate.

Jazeem replied:
Changed this

Jazeem commented on open_spiel/games/2048.h:
```diff
@@ -0,0 +1,159 @@
+namespace two_zero_four_eight {
+
+constexpr int kNumPlayers = 1;
+constexpr int kDefaultRows = 4;
```
Changed this

Jazeem commented on open_spiel/games/2048.h:
```diff
@@ -0,0 +1,159 @@
+};
+
+// Game object.
+class TwoZeroFourEightGame : public Game {
```
Changed this

Jazeem commented on open_spiel/games/2048.h:
```diff
@@ -0,0 +1,159 @@
+};
+
+// State of an in-play game.
+class TwoZeroFourEightState : public State {
```
Changed this

Jazeem commented on open_spiel/games/2048.h:
```diff
@@ -0,0 +1,159 @@
+#include "open_spiel/spiel.h"
+
+namespace open_spiel {
+namespace two_zero_four_eight {
```
Changed this

Jazeem commented on open_spiel/games/2048.h:
```diff
@@ -0,0 +1,159 @@
+constexpr int kMaxScore = INT_MAX;
+
+// The chance tiles that randomly appear on the board after each move
+const std::vector<int> kChanceTiles = {2, 4};
```
Changed this

Jazeem commented on open_spiel/games/2048.cc:
```diff
@@ -0,0 +1,382 @@
+    r += direction_diff.row;
+    c += direction_diff.column;
+  } while (WithinBounds(r, c) && CellAvailable(r, c));
+  return std::vector<Coordinate> {prev,
```
Fixed this

Jazeem commented on open_spiel/games/2048.cc:
```diff
@@ -0,0 +1,382 @@
+void TwoZeroFourEightState::DoApplyAction(Action action) {
+  if (IsChanceNode()) {
+    // The original 2048 game starts with two random tiles
+    if (!extra_chance_turn_) {
```
I think I'd have to duplicate the chance player action logic in ChanceOutcomes() and DoApplyAction() to accomplish this in game initialisation. Am I missing some other way?

Jazeem commented on open_spiel/games/2048.cc:
```diff
@@ -0,0 +1,382 @@
+      return "No Cell Available";
+    }
+    ChanceAction chance_action = SpielActionToChanceAction(action_id);
+    return absl::StrCat(std::to_string(chance_action.is_four ? 4 : 2), 
```
Fixed this

Jazeem commented on open_spiel/games/2048.cc:
```diff
@@ -0,0 +1,382 @@
+}
+
+void TwoZeroFourEightState::UndoAction(Player player, Action action) {  
+  history_.pop_back();
```
I was thinking of not implementing UndoAction if it's not necessary for all games in OpenSpiel. I can remove this function if that's the case.

Otherwise I think I'll just have to keep a copy of the previous state as reconstructing the state from the move played and other associated info can get quite complex

Jazeem commented on open_spiel/games/2048.cc:
```diff
@@ -0,0 +1,382 @@
+void TwoZeroFourEightState::PrepareTiles() {
+  for (int r = 0; r < kDefaultRows; r++) {
+    for (int c = 0; c < kDefaultColumns; c++) {
+      Tile tile = BoardAt(r, c);
```
I think c++ won't allow assignments to function returns like this. 
But I have introduced a function for this and made this code a bit more simpler

Jazeem commented on open_spiel/games/2048.cc:
```diff
@@ -0,0 +1,382 @@
+  return BoardAt(r, c).value == 0;
+}
+
+Coordinate GetVector(int direction) {
```
Can you provide a bit more info?

Jazeem commented on open_spiel/games/2048.cc:
```diff
@@ -0,0 +1,382 @@
+      action_bases, {move.row, move.column, move.is_four});
+}
+
+std::vector<std::vector<int>> TwoZeroFourEightState
```
I've made one of these an array as to not hardcode 4 here since there's kRows and KColumns

tewalds commented on open_spiel/games/2048.cc:
```diff
@@ -0,0 +1,378 @@
+    return {};
+  }
+  if (IsChanceNode()) {
+    return LegalChanceOutcomes();
```
LegalChanceOutcomes doesn't seem to be defined anywhere. Is this ChanceOutcomes()?

Jazeem replied:
I was referencing backgammon.cc for writing this.
As I understand, LegalChanceOutcomes() is defined in spiel.h and it uses the output of ChanceOutcomes() to return vector<Action>

Jazeem commented on open_spiel/games/2048.cc:
```diff
@@ -0,0 +1,378 @@
+                         GameType::ChanceMode::kExplicitStochastic,
+                         GameType::Information::kPerfectInformation,
+                         GameType::Utility::kGeneralSum,
+                         GameType::RewardModel::kTerminal,
```
My bad. Forgot to change this, initially the code was written as a win/loss game. Fixed it now.

Jazeem commented on open_spiel/games/2048.cc:
```diff
@@ -0,0 +1,378 @@
+constexpr int kMoveRight = 1;
+constexpr int kMoveDown = 2;
+constexpr int kMoveLeft = 3;
+inline const std::vector<Action> kPlayerActions() {
```
LegalActions() expects a vector as returns. That's why kept this as vector too.

Jazeem commented on open_spiel/games/2048.cc:
```diff
@@ -0,0 +1,378 @@
+      reverse(y.begin(), y.end());
+      break;
+    case kMoveLeft:
+    case kMoveDown:
```
True. Simplified this.

lanctot commented:
Hi @Jazeem, something changes on the MacOS behavior via GitHub Actions, can you apply these changes to trigger the tests again: https://github.com/deepmind/open_spiel/pull/907/files

lanctot commented:
Yeah it'll take more than just that one line, now there seems to be issues with the Jax versions .. :( 

Keep an eye on https://github.com/deepmind/open_spiel/pull/907, might only be resolved tomorrow.

Jazeem commented:
> Keep an eye on https://github.com/deepmind/open_spiel/pull/907, might only be resolved tomorrow.

Sure. Will do 👍

lanctot commented:
> > Keep an eye on #907, might only be resolved tomorrow.
> 
> Sure. Will do 👍

Ok tests are passing, so you just need to update the Jax and etc. package versions (one line change in install.sh)

lanctot commented:
@Jazeem thanks, looking good! Are you done fixing things up for the moment?

I've contacted @tewalds who will review the import, would be good to pull this one in now if you're done making changes.

Jazeem commented:
@lanctot 
Yes, done fixing things up for now.

lanctot commented:
Hi @Jazeem, it's now been merged, but a fair bit changed from internal review. Please take a look through it let us know what you think. You can still do the almost-infinite version by setting max_tile to a very high value.

Jazeem commented:
@lanctot It looks great 👍

Only thing I noticed is that if a player/agent makes a move that doesn't change anything on the board, then additional chance tiles are not added. So there could be a few inconsequential moves like that, so not sure if those should be counted in

`int MaxGameLength() const override { return 2 * 2 * max_tile_; }`
